Added cityscapes dataset

f1recracker opened the pull request:
Added Cityscapes dataset with segmentation labels. Fixes #214.

f1recracker commented:
Hello! This is my first contribution to tensorflow-datasets! I actually had a few quick questions / design items to discuss before merging this.

1. How are the unit tests run? (is it via `oss_scripts/oss_tests`?) I haven't run any of the cityscapes test but I checked the size of using the actual dataset in my local repository.

2. What is the recommended size of fake datasets used for testing? I have tried to ensure each split has a different size but I still end up with a 38MB fake dataset (cityscapes has big images - 1024 * 2048).

f1recracker commented:
I've used pngquant to compress the train images!

us commented:
@f1recracker some dataset's test generate own images and label like `testing/cifar.py`. Maybe it can reduce size even more.

cyfra commented:
@f1recracker - for running the tests, you can also invoke them from command line directly (assuming you're in the top directory):

python3 -m tensorflow_datasets.image.celeba_test

cyfra commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,185 @@
+
+    if any(not os.path.exists(p) for p in paths.values()):
+      msg = 'You must download the dataset files manually and place them in: '
+      msg += ', '.join(paths.values())
```
can you add some explanation on why the tfds cannot download the files?

f1recracker replied:
Cityscapes (much like ImageNet) requires a login before you can download the files.

cyfra replied:
:-(  I see. please mention it in the comment here.

cyfra commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,185 @@
+# pylint: disable=bad-indentation
```
any reason why ? ;-)

f1recracker replied:
Sorry! My global pylint config uses 4 spaces which causes a lot of false positives in Google style python code! Will remove!

cyfra commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,185 @@
+class CityscapesCore(tfds.core.GeneratorBasedBuilder):
+  '''Base for cityscapes datasets'''
+
+  VERSION = tfds.core.Version('2.0.0')
```
the 2.0 - is it on purpose? (by default, we start with 1.0 versions for the new dataset)

f1recracker replied:
No I just used the latest I could find from other examples. Will move to 1.0

cyfra commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,185 @@
+        urls=["https://www.cityscapes-dataset.com", "https://github.com/mcordts/cityscapesScripts"],
+        citation=_CITATION,
+    )
+
```
Add @abc.abstractmethod

cyfra commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,185 @@
+_DESCRIPTION = '''\
+  Cityscapes is a dataset consisting of diverse urban street scenes and dense pixel-level
+  annotations taken across 50 different cities at varying times of the year. The label
+  annotations span across 30 (34 including ingore indices) distinct classes of objects
```
typo 'ingore -> ignored' ?

cyfra commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,185 @@
+  https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/helpers/labels.py#L52-L99
+
+  The dataset consists of 5000 annotated image, label pairs at 1024 * 2048 resolution
+  pre-split into training (2975) validation (500) and test (1525) sets.
```
Please add:

WARNING: this dataset requires users to setup a login and password in order to get the files.

f1recracker replied:
Where would this go? end of the first paragraph? end of description string?

cyfra replied:
Wherever you prefer. 
You could add it at the end.

Later it will be automatically copied to datasets.md

cyfra commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,185 @@
+IMAGE_FILE_RE = re.compile(r'([a-z\-]+)_(\d+)_(\d+)_leftImg8bit\.png')
+
+def _image_id(image_file):
+  match = IMAGE_FILE_RE.match(image_file)
```
add comment on what this function does.

cyfra commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,185 @@
+    paths['labels'] = os.path.join(dl_manager.extract(paths['labels']), 'gtFine')
+
+    # Number of shards calculations:
+    # Max shard size = 4GB
```
very nice - thanks.

cyfra commented on tensorflow_datasets/image/cityscapes_test.py:
```diff
@@ -0,0 +1,27 @@
+      'test': 2,
+  }
+
+  DATASET_CLASS = cityscapes.CityscapesCoarse
```
does it work? I think that you might need to create another class in this file (otherwise you might be just overwriting the entry from line 12)

f1recracker replied:
Yes this is incorrect! I wasn't sure how to run the unit tests until today and I need another class.

f1recracker commented:
I've fixed most of the changes but I have observed that `dl_manager.extract` does not work as intended in unit tests (which actually returns DL_EXTRACT_RESULT or the example_path from what I can see here ) but `dl_manager.iter_archive` works. Is this intentional? I'm actually rewriting my code using `dl_manager.iter_archive`.

cyfra commented:
@f1recracker - yes, you're right, this is an inconsistency in our testing framework that we introduced when adding DL_EXTRACT_RESULT.


@pierrot0 - Pierre, can you comment on the extract issue and what is the recommended way @f1recracker should follow?

f1recracker commented:
Has there been any updates on this? Sorry I got a little caught up with some work during the past few weeks!

f1recracker commented:
It seems that the latest version fixed the inconsistency and also another error. I've also refactored my code to now use BuilderConfigs instead of having separate classes for Cityscapes - fine and Cityscapes - coarse.

lgeiger commented:
@f1recracker Thanks for the PR! Would it be also possible to add the depth data of the Cityscapes dataset as an additional feature?

f1recracker commented:
I wasn't aware of the depth maps - let me have a look at this.

lgeiger commented:
Thanks for taking a look!

lgeiger commented:
@f1recracker The stereo pair right images would also be a great addition to this PR.

f1recracker commented:
@lgeiger Added support for right images and disparity maps too. Cityscapes now has 4 configurations:
- 'semantic_segmentation' - left + segmentations
- 'semantic_segmentation_extra' - left + segmentations + coarse labels + train extra split
- 'stereo_disparity' - left + right + disparity maps
- 'stereo_disparity_extra' - left + right + disparity maps + train extra split

Could this be reviewed again? There's significant changes since the last review (BuilderConfigs so no abcs, and the recent change with disparity maps)? Thanks!

us commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -194,24 +207,24 @@ def calculate_num_shards(split_size, features_size_mb):
-      ),
+        tfds.core.SplitGenerator(
+            name=tfds.Split.TRAIN,
+            num_shards=calculate_num_shards(2975, features_size_mb),
```
You don't need to specify num_shards anymore.

f1recracker replied:
Wasn't aware of this - thanks!

netw0rkf10w commented:
@f1recracker Thanks for working on this.
@cyfra What would need to be changed for this to be merged please?

cyfra commented:
Hey @netw0rkf10w 
Unfortunately we have grown quite a backlog of PRs - sorry for that. 
We were quite busy with making sure that current ones will work with Python3.

as I see that there is high interest with this dataset, I'll try to merge it soon.

netw0rkf10w commented:
@cyfra Thank you for your reply, and for your hard work! This is indeed a very popular dataset in computer vision, so the merge would benefit a lot of users. Looking forward to having this merged.

Conchylicultor commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,280 @@
+        if self.builder_config.segmentation_labels:
+          features['segmentation_label'] = os.path.join(
+              paths_city_root['segmentation_labels'],
+              f'{image_id}_{self.builder_config.label_suffix}.png')
```
This is using python3 only feature, while TFDS should be both Py2 and 3 compatible.

hermannsblum replied:
@f1recracker https://github.com/f1recracker/tensorflow-datasets/pull/1 here is a possible fix

f1recracker replied:
will fix tonight. Thanks

f1recracker commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,280 @@
+
+  BUILDER_CONFIGS = [
+      CityscapesConfig(
+          name='semantic_segmentation',
```
It was a deliberate choice - Disparity maps need additional permissions from the authors and aren't readily available. 'Extra' splits - whether for segmentation or for the disparity requires additional files (50GB). I chose this route so that each configuration so that it would work for all cases.

f1recracker replied:
I gave this some more thought today to see if I could refactor the code to use fewer if/else statements. The problem arises because Cityscapes has too many features that every user may not use.

1. There's two 'datasets' for different tasks - semantic semgentation and disparity inference. Disparity tasks can also require the right images - another additional download.
2. Every cityscapes 'dataset' (semantic seg / stereo depth) has an additional 'extra' split. This split is much larger and any labels provided are coarse grain instead.

I think we still need 4 (2 (task) x 2 (extra split)) configurations.

One feature that might reduce some if/else might be to use a constant feature dict and use a 'default' image tensor (since cityscape labels are images) if this is acceptable.

Aya-S replied:
Are there any tutorials on how to use this class please?

f1recracker replied:
You just have to do something along the lines of:

```
import tensorflow_datasets as tfds
dataset = tfds.load(name="cityscapes/semantic_segmentation_extra", split="train")
# Can also specify config from above.
```

I'd give a working Colab but unfortunately this is a pretty big dataset and requires manual download :)

Aya-S replied:
@f1recracker thank you for the reply! 
I downloaded the dataset manually and will attempt to access it with the syntax you suggested
Thanks again.

f1recracker commented on tensorflow_datasets/image/cityscapes.py:
```diff
@@ -0,0 +1,280 @@
+    # Enable features as necessary
+    features = {}
+    features['image_id'] = tfds.features.Text()
+    features['image_left'] =\
```
will fix tonight. Thanks.

netw0rkf10w commented:
Thanks a lot, @f1recracker, for continuing working on this! Unfortunately I am not yet familiar enough with the code to be able to help you.

cyfra commented:
Ok - dataset is submitted. 

Thanks a lot @f1recracker for your contribution and patience.
